Added Angle/RightAngle classes for two intersecting lines

PhotonSpheres opened the pull request:
<!--
Thank you for contributing to manim!

Please ensure that your pull request works with the latest
version of manim from this repository.
-->

## Motivation
I feel like simple angles of two lines will be needed often enough to implement them permanently.

## Overview / Explanation for Changes
I added two classes: `ArcAngle` and `RightAngle`. They both take two lines as input. Then one can specify what possible angle they want to draw and customize the look. Here is a small video introducing the possibilities:

https://user-images.githubusercontent.com/44864613/110237913-e6b12100-7f3e-11eb-9437-3a3dbebbe2dc.mp4

## Oneline Summary of Changes
<!-- Please update the lines below with a oneline summary
for your changes. It will be included in the list of upcoming changes at
https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release -->
```
- Added feature: Angles of intersecting lines
```

## Testing Status
<!-- Optional (but recommended): your computer specs and
what tests you ran with their results, if any. This section
is also intended for other testing-related comments. -->
I ran pytest on my MacBook Pro with macOS Catalina 10.15.17 with `194 passed in 94.57s (0:01:34)`

## Further Comments
<!-- Optional, any further comments regarding your PR
that might be useful for reviewers.. -->
This is my first PR and I am quite new to git/GitHub. Feel free to show me everything I have done wrong and I will try my best to change it.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

<!-- Once again, thanks for helping out by contributing to manim! -->


<!-- Do not modify the lines below. -->
## Reviewer Checklist
- [x] Newly added functions/classes are either private or have a docstring
- [x] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [x] The title is descriptive enough

jsonvillanueva commented:
Hey, thanks for the first time contribution! This looks like a good feature to have in Manim's core; I'll submit a review in a bit.

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,254 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+                )
+                self.wait()
+
+    ..manim:: ArcAngleExample
+
+        class Angles_Test3(Scene):
```
The class name don't match and so this fails to render in[ your PR's documentation](https://manimce--1085.org.readthedocs.build/en/1085/reference/manim.mobject.geometry.ArcAngle.html#manim.mobject.geometry.ArcAngle). Also, this should be made an image.
```suggestion
    .. manim:: Angles_Test3
        :save_last_frame:

        class Angles_Test3(Scene):
```

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,254 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+    dot_distance : :class:`float`
+        Placement of the dot in the arc. This distance from the arc center is given as the radius
+        diveded by this :class:`float` parameter.
+    dot_color : :class:`Colors`
```
```suggestion
    dot_color : :class:`~.Colors`
```

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,254 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+    Examples
+    --------
+
+    .. manim:: RightAngleExample
+
+        class RightAngleExample(Scene):
```
```suggestion
    ..manim:: RightAngleExample
        :save_last_frame:

        class RightAngleExample(Scene):
```

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,254 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+                    line_list[k].add( lineb )
+                    line_list[k].add( rightangles[k] )
+                line_list.arrange_in_grid(buff=1.5)
+                self.play(
+                    *[ShowCreation(
+                        line_list[k][:2]
+                    ) for k in range(4)]
+                )
```
Because this animation doesn't need to be a video, we could change these (and the other examples) to just use `self.add(....)` instead of using animations (i.e. `self.play(...)`, `self.wait(...)`. This will allow the documentation to be generated more quickly.

PhotonSpheres replied:
I changed that, thanks!

jsonvillanueva reviewed (changes requested):
In general, this looks good as a first pass. I haven't gotten to use the changes yet, but I'll get to that momentarily. Also, you can generate the documentation locally by navigating to `path/to/manim/docs` and running `make html`.

WampyCakes commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,257 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+
+
+class RightAngle(VMobject):
+    """Am elbow-type like object representing a right angle of two intersecting lines.
```
```suggestion
"""An elbow-type mobject representing a right angle between two intersecting lines.
```

WampyCakes commented:
I think this will be a great addition, but I have three questions/comments.
1. Is it absolutely necessary that this be 2 classes? I'd prefer an angle marker that sees what the angle between two lines is and then puts the correct one.
2. If it is necessary, there is an `Elbow` class in manim. While you'd be doing the same thing by overriding `__init__`, I think it would be a little clearer to someone browsing the code that this is just an elbow if the parent class was `Elbow` rather than `VMobject`. Thoughts on this?
3. I haven't tried this, so I'm going to ask. What would happen if you used more than 360 or 720 degrees? Can it do something [like this](https://i.stack.imgur.com/SKUkJ.png)?

PhotonSpheres commented:
> I think this will be a great addition, but I have four questions/comments.

Thanks for the comments!

> 1. Is it absolutely necessary that this be 2 classes? I'd prefer an angle marker that sees what the angle between two lines is and then puts the correct one.

I guess what would be the right one? Do you always want to mark your right angles as an elbow or as an arc with a dot? Or sometimes mark a right angle as an arc without a dot? This is why I put it more open like that.

> 2. If it is necessary, there is an `Elbow` class in manim. While you'd be doing the same thing by overriding `__init__`, I think it would be a little clearer to someone browsing the code that this is just an elbow if the parent class was `Elbow` rather than `VMobject`. Thoughts on this?

I found it cleaner looking like that but I understand the sentiment. Will think about how it would look if `RightAngle` would inherit from `Elbow`.

> 3. I haven't tried this, so I'm going to ask. What would happen if you used more than 360 or 720 degrees? Can it do something [like this](https://i.stack.imgur.com/SKUkJ.png)?

It cannot since the class calculates the angle on its own and will always come up with a number between 0 and 2pi.

> 4. Lastly, this isn't majorly important, but the lines don't necessarily have to intersect; they only have to share a common vertex. Maybe update the documentation to reflect that it merely creates an angle marker between two lines?

Not sure what you mean here. I phrased it like that since the intersection is calculated by a predefined calculator. This already has an error function like "Lines do not intersect" if one uses parallel lines as an input.

WampyCakes commented:
> I guess what would be the right one? Do you always want to mark your right angles as an elbow or as an arc with a dot? Or sometimes mark a right angle as an arc without a dot? This is why I put it more open like that.

I hadn't thought about it quite like that, but you could morph them into one class and have a parameter for which to use on right angles. Like a parameter `use_elbow_for_right_angle=True`, if false use an arc. I'm just more in favor of having one type of object that handles two scenarios than two objects that each handle one scenario; it requires less choice by the end user also. Because the same class could be applied to both types of angles, if someone has a loop where they're generating many different angle markers, they won't have to include their own logic to handle right angles if they have right angles. Rather they could just use one class for all of them.

> Not sure what you mean here. I phrased it like that since the intersection is calculated by a predefined calculator. This already has an error function like "Lines do not intersect" if one uses parallel lines as an input.

[Take a look at this picture.](https://en.wikipedia.org/wiki/Angle#/media/File:Two_rays_and_one_vertex.png) I don't think about the two rays making up this angle as intersecting. It can be the case where you're drawing angle markers between intersecting lines (like your example videos), but I don't want users to be reading through the documentation and dismiss angle markers because it says "intersecting lines" and they think it won't apply then. I'm not saying that this reasoning is legitimate because the examples are clear, but it seems plausible to happen when you think about the sometimes unpredictable behavior of end users. It's a minor nitpick, but I don't think the intersecting line language is necessary.

PhotonSpheres commented:
> I hadn't thought about it quite like that, but you could morph them into one class and have a parameter for which to use on right angles. Like a parameter `use_elbow_for_right_angle=True`, if false use an arc. I'm just more in favor of having one type of object that handles two scenarios than two objects that each handle one scenario; it requires less choice by the end user also. Because the same class could be applied to both types of angles, if someone has a loop where they're generating many different angle markers, they won't have to include their own logic to handle right angles if they have right angles. Rather they could just use one class for all of them.

Ah I get it. My reasoning was more about keeping it simple and be able to animate it. I wouldn't want the angle to have a dot for one frame if an animation happens to cross a 90° mark. It is more like a simple function to draw an angle. Also, in a sketch one still should be able to draw a right angle for something that isn't technically a right angle. Hence the `dot=True` parameter is a good toggle IMHO.

> [Take a look at this picture.](https://en.wikipedia.org/wiki/Angle#/media/File:Two_rays_and_one_vertex.png) I don't think about the two rays making up this angle as intersecting. It can be the case where you're drawing angle markers between intersecting lines (like your example videos), but I don't want users to be reading through the documentation and dismiss angle markers because it says "intersecting lines" and they think it won't apply then. I'm not saying that this reasoning is legitimate because the examples are clear, but it seems plausible to happen when you think about the sometimes unpredictable behavior of end users. It's a minor nitpick, but I don't think the intersecting line language is necessary.

Makes sense! Although I guess this would technically still be an intersection. Since the class uses the `line_intersection()` function in the code I would imagine this wording is fine as to not confuse the people when trying to understand how it works.

WampyCakes commented:
That's a fair reason for the first point.

> Makes sense! Although I guess this would technically still be an intersection. Since the class uses the `line_intersection()` function in the code I would imagine this wording is fine as to not confuse the people when trying to understand how it works.

Many users will never look at the underlying code, they just want to look at the docs and see what's possible and how. I don't think that someone who looks at the code and understands it would be thrown by the `line_intersection()` function. If they generally get how the method should work, it will make sense. If not, those are the people who are benefited by removing mention of intersecting lines in the docs.

MrMallIronmaker commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,221 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+        self,
+        line1,
+        line2,
+        radius=0.3,
+        quadrant=(1, 1),
+        other_angle=False,
+        dot=False,
+        dot_radius=0.03,
+        dot_distance=1.8,
+        dot_color=WHITE,
```
Good job in setting meaningful defaults - quadrant especially.

PhotonSpheres replied:
Cheers! I changed `dot_radius`to `None` and the default will be calculated as 1/10 of the arc radius. Those two sizes should be related most of the time.

MrMallIronmaker commented:
Welcome to Github / Manim dev / open-source development! This is a great contribution, and an especially excellent first contribution.

I like @WampyCakes' mention of angles greater than 360, but it makes more sense to leave it for future work. Same with potential for double- or triple-struck arcs.

I've also left some comments on the `quadrant` and `dot_distance` parameters.

PhotonSpheres commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,221 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+    radius : :class:`float`
+        The radius of the :class:`Arc`.
+    quadrant : Sequence[:class:`int`]
+        A sequence of two :class:`int` numbers determining which of the 4 quadrants should be used.
```
This does make sense! I added this exact wording.

PhotonSpheres commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,221 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+        The radius of the :class:`Dot`.
+    dot_distance : :class:`float`
+        Placement of the dot in the arc. This distance from the arc center is given as the radius
+        diveded by this :class:`float` parameter.
```
> The value itself could range from 0 (right at the center) to 1 (right on the arc).

I did implement that, a great idea! This is now a linear increase from the intersection to the arc in terms of radius, hence 0 is the intersection, and 1 the arc. Does `dot_distance` make enough sense in this context now you think?

PhotonSpheres commented:
> Welcome to Github / Manim dev / open-source development! This is a great contribution, and an especially excellent first contribution.

Cheers!

> I like @WampyCakes' mention of angles greater than 360, but it makes more sense to leave it for future work. Same with potential for double- or triple-struck arcs.

Yeah, sounds like a fun addition later on.

> I've also left some comments on the `quadrant` and `dot_distance` parameters.

I implemented your ideas, thanks for that! I also changed the `dot_radius` to be 1/10 of the arc radius if not specified otherwise.

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,228 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+        self,
+        line1,
+        line2,
+        radius=0.3,
```
```suggestion
        radius=None,
```
I'd prefer this default to be `None` and calculated dynamically based on the size of the shorter of the two segments from the intersection. See this example:
```py
class Test(Scene):
    def construct(self):
        l1 = Line(ORIGIN, RIGHT)
        l2 = Line(ORIGIN, UP*.2)
        a = ArcAngle(l1, l2)
        self.add(l1,l2,a)
```
![image](https://user-images.githubusercontent.com/29168469/110840115-32bed700-8259-11eb-9c5c-d7e040b19911.png)

MrMallIronmaker replied:
Would you say something like the smaller of 75% of the shorter side or 0.3? Or would two large lines produce a large angle by default?

jsonvillanueva replied:
I think 75%, 50%, or 66.66666% are sensible for the shorter side. I also think two long line subsections should produce a large angle (radius from the "intersection", not theta) by default. Perhaps there can be hard cap (i.e. flooring) for the radius size as well. It's just awkward that the default radius will sometimes not hit the line(s).

PhotonSpheres replied:
I did implement that with a hard cap of 0.4 as the biggest radius. Other than that the radius will be 2/3 of the shorter line. Thanks for the idea!

jsonvillanueva reviewed (commented):
Sorry it's taken me a while to get back to this. I've left a few comments, all asking about non-blocking changes since this functions mainly as expected.

PhotonSpheres commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,228 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+
+
+class RightAngle(VMobject):
+    """An elbow-type mobject representing a right angle between two lines.
```
I changed to inherit from both, `Arc` and `Elbow`, but my naive code isn't working. It has something to do with multiple inheritances, that is I get `AttributeError: Angle object has no attribute 'start_angle'`when running an `elbow=True` script. I have no experience with that and I don't know how easy of a fix that is. Any ideas?

jsonvillanueva replied:
This has to do with the [method resolution order](http://www.srikanthtechnologies.com/blog/python/mro.aspx) in multiple inheritance... what's happening is Elbow is executing `__init__` which calls the parent classes' `__init__` up the hierarchy to Mobject. At this point in `Mobject.__init__` makes a call to [generate_points()](https://github.com/PhotonSpheres/manim/blob/b56c6949e8c8a7fd500138b90ec4fca17337b51f/manim/mobject/mobject.py#L61) and since Arc comes first in the multiple inheritance, it uses `Arc.generate_points()` which fails since Elbow doesn't have `self.start_angle` (but Arc does).

A couple of solutions in order of preference:
1. Add Elbow.generate_points() (moving most of Elbow's `__init__()` to `generate_points`), and add `Angle.generate_points` to call the correct generate_points:
```py
class Angle(Arc, Elbow):
# ....
    def generate_points(self):
        if self.elbow:
            Elbow.generate_points(self)
        else:
            Arc.generate_points(self)
```
^ Note: there's likely a better/more "pythonic" method than this. If anyone else has ideas, feel free to chime in as there's not many examples of multiple inheritance in this library.

2. Inherit from VMobject in Angle instead (not recommended). This will serve to call VMobject.generate_points when doing Elbow.__init__(), but at the cost of losing the dual inheritance.

PhotonSpheres replied:
Thanks for the help! Unfortunately, I am not really sure how to implement your first idea. Do you have a small working example of such a script I could look at?

jsonvillanueva replied:
Absolutely, here's a [quick commit](https://github.com/jsonvillanueva/manim/commit/7844569fc7390da5df3a543ad662f4a04b3ff879) off your latest changes. It seems you don't actually need to split the logic off of Elbow into `generate_points()`. It's enough to just override it in `Angle`.

jsonvillanueva replied:
Also, there should be some test cases for this PR -- just in the event that something in the future changes we know if any expected behavior breaks/changes. Read up on [this section](https://github.com/ManimCommunity/manim/wiki/Testing#graphical-unit-test) on graphical unit testing which involves creating control data (.npz file) for your new Angle object. I think this will complete the PR (in my eyes at least).

PhotonSpheres replied:
Oh, that is very neat, cheers!

PhotonSpheres replied:
I added GU tests for `Angle` and `RightAngle`. I am not sure how complicated they should be and I only added the angles to the screen, not the lines. Would that be a good idea?

jsonvillanueva replied:
Adding only the angles is fine since that's ultimately what's being tested, but maybe an additional test case with the lines is necessary. I do believe we could use increased complexity and really get at some of the edge cases (like the case I brought up about shorter line segments/dynamic sizing of the radius). I think your examples for the docs would also serve as a good test (a multitude of angles being tested all at the same time in one frame).

Speaking of your docs... they didn't render. Will have to look into that. Are they rendering locally for you?

jsonvillanueva commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,236 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+    """
+
+    def __init__(self, line1, line2, length=None, **kwargs):
+        if length == None:
+            Angle.__init__(self, line1, line2, elbow=True, **kwargs)
+        else:
+            Angle.__init__(self, line1, line2, radius=length, elbow=True, **kwargs)
```
I think this can be simplified to just:
```suggestion
        Angle.__init__(self, line1, line2, radius=length, elbow=True, **kwargs)
```
since if `length==None` then the Angle constructor would just receive radius=None and calculate it dynamically.

jsonvillanueva reviewed (approved):
Thanks again for the contribution, LGTM!

hydrobeam commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+
+    Examples
+    --------
+    The first example shows some right angles with a dot in the middle while the second example shows
```
Isn't it a bit redundant to have an example for RightArcAngle  here and also in its own separate class? The sister class can be mentioned in a `See Also` section and the example for the RightArcAngle left only under the documentation for RightAngle.

PhotonSpheres replied:
This example shows arc-like right angles, that is, an arc with a dot in the middle. The class `RightAngle` gives elbow-like indicators for right angles. So those examples just show that both conventions can be used in manim. Does this make sense?

hydrobeam replied:
That does seem reasonable, my bad.

PhotonSpheres replied:
All good!

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+        anchor_angle_1 = inter + quadrant[0] * radius * line1.get_unit_vector()
+        anchor_angle_2 = inter + quadrant[1] * radius * line2.get_unit_vector()
+
+        if elbow == True:
```
```suggestion
        if elbow:
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+            [line1.get_start(), line1.get_end()], [line2.get_start(), line2.get_end()]
+        )
+
+        if radius == None:
```
```suggestion
        if radius is None:
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+            angle_1 = angle_of_vector(anchor_angle_1 - inter)
+            angle_2 = angle_of_vector(anchor_angle_2 - inter)
+
+            if other_angle == False:
```
```suggestion
            if not other_angle:
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+                arc_center=inter,
+                **kwargs
+            )
+            if dot == True:
```
```suggestion
            if dot:
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+                **kwargs
+            )
+            if dot == True:
+                if dot_radius == None:
```
```suggestion
                if dot_radius is None:
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+    elbow : :class:`bool`
+        Produces an elbow-type mobject indicating a right angle, see :class:`RightAngle` for more information
+        and a shorthand.
+    kwargs
```
```suggestion
    **kwargs
```

friedkeenan commented on manim/mobject/geometry.py:
```diff
@@ -1554,3 +1556,243 @@ def __init__(self, main_shape, *mobjects, **kwargs):
+        The second line.
+    length : :class:`float`
+        The length of the arms.
+    kwargs
```
```suggestion
    **kwargs
```

friedkeenan reviewed (commented):
Looks fine to me, though I will admit that the multiple inheritance is a *tad* odd, but ultimately I'm fine with it. Just have a few nitpicks

PhotonSpheres commented:
> Looks fine to me, though I will admit that the multiple inheritance is a _tad_ odd, but ultimately I'm fine with it. Just have a few nitpicks

Cool, thanks for your time reviewing it and for the improvements!

behackl reviewed (approved):
An excellent addition, and fabulous documentation. This is a perfect example of how I would like most of Manim's classes to be documented at some point.

Thank you very much for your contribution!

kolibril13 commented:
Thanks a lot from me as well!
I just had the idea of adding another example (in this case with a moving line, close to the example that you posted on Twitter about the deg/Fahrenheit conversion)
What do you think about this scene into the example gallery:
```python
class Test(Scene):
    def construct(self):
        theta_tracker = ValueTracker(45)
        line1 = Line(LEFT * 0.1, RIGHT)
        line_moving = Line(LEFT * 0.1, RIGHT).rotate_about_origin(
            theta_tracker.get_value() * DEGREES
        )
        line2 = line_moving.copy()
        a = Angle(line1, line_moving, radius=0.5, other_angle=False)
        te = MathTex(r"\theta").next_to(a, RIGHT, buff=SMALL_BUFF)
        self.add(te)
        line_moving.add_updater(
            lambda x: x.become(line2.copy()).rotate_about_origin(
                theta_tracker.get_value() * DEGREES
            )
        )
        a.add_updater(
            lambda x: x.become(Angle(line1, line_moving, radius=0.5, other_angle=False))
        )
        self.add(line1, line_moving, a)
        self.play(theta_tracker.animate.increment_value(81))
```

https://user-images.githubusercontent.com/44469195/113064800-080cd200-91b8-11eb-8d94-1f36aff5715d.mp4

WampyCakes commented:
@kolibril13 I think it would look better if the end of both lines had a common vertex rather than extending through each other 🤷‍♂️
